### PR TITLE
Make TweakScore and CustomScore mutable at the segment level

### PR DESCRIPTION
**Pull request by robyoung**
Addresses issue #806

I have also added a test that shows using `tweak_score` with a facet. This is the problem I was originally describing when I raised this on Gitter. I'm not sure if this kind of test is wanted or whether this is the correct place for it. Would love some guidance on that.

### CODE REVIEW

**Comment by robyoung** [comment]
Is there anything else I can do to this PR to make it easier to review?

**Comment by fulmicoton** [comment]
@robyoung nothing. I am usually much more reactive, but due to COVID life has been a little hectic recently. Please wait a little bit longer!

File: src/collector/custom_score_top_collector.rs [hunk 1]
```diff
@@ -117,9 +117,9 @@ where
 
 impl<F, TScore> CustomSegmentScorer<TScore> for F
 where
-    F: 'static + Sync + Send + Fn(DocId) -> TScore,
+    F: 'static + Sync + Send + FnMut(DocId) -> TScore,
```
Comment (fulmicoton) [inline src/collector/custom_score_top_collector.rs]: I haven't tried it but can we remove `Sync` here too?

File: src/collector/tweak_score_top_collector.rs [hunk 2]
```diff
@@ -121,9 +121,112 @@ where
...
+                    let query_ords: HashSet<u64> = facets
+                        .iter()
+                        .filter_map(|key| facet_dict.term_ord(key.encoded_str()))
+                        .collect();
+
+                    let mut facet_ords_buffer: Vec<u64> = Vec::with_capacity(20);
+
+                    move |doc: DocId, original_score: Score| {
+                        ingredient_reader.facet_ords(doc, &mut facet_ords_buffer);
+                        let missing_ingredients = facet_ords_buffer
```
Comment (fulmicoton) [inline src/collector/tweak_score_top_collector.rs]: That seems very suboptimal... What about just iter().filter().count()?

File: src/collector/tweak_score_top_collector.rs [hunk 3]
```diff
@@ -121,9 +121,112 @@ where
...
+#[cfg(test)]
+mod tests {
+    use crate::collector::TopDocs;
+    use crate::query::BooleanQuery;
+    use crate::schema::*;
+    use crate::{DocId, Index, Score, SegmentReader};
+    use std::collections::HashSet;
+
+    #[test]
+    fn tweak_score_with_facets() {
```
Comment (fulmicoton) [inline src/collector/tweak_score_top_collector.rs]: This looks more like an example than a test. Could you move this to an example?

File: src/collector/custom_score_top_collector.rs (hunk 1 above)
Comment (fulmicoton) [reply src/collector/custom_score_top_collector.rs]: ... and possibly Send.

File: src/collector/tweak_score_top_collector.rs (hunk 3 above)
Comment (robyoung) [reply src/collector/tweak_score_top_collector.rs]: Absolutely! Great idea, I knew it felt wrong here.

**Comment by robyoung** [comment]
No worries at all, I totally understand. I have an 11 month old, COVID makes life hard. Thank you for the review, I'll take a look at addressing them later today.

File: src/collector/custom_score_top_collector.rs (hunk 1 above)
Comment (robyoung) [reply src/collector/custom_score_top_collector.rs]: I don't really understand the ramifications of this but the tests pass.

File: src/collector/tweak_score_top_collector.rs (hunk 2 above)
Comment (robyoung) [reply src/collector/tweak_score_top_collector.rs]: Good point, this was here for debugging but when I was figuring things out but doesn't need to be any more.

**Comment by fulmicoton** [comment]
Awesome you even thought about changing the custom scorer! merging.

**Pull request merged**